staging/src/k8s.io/kube-aggregator/pkg/controllers/openapiv3/aggregator/aggregator.go

```diff
@@ -98,109 +86,123 @@ func BuildAndRegisterAggregator(downloader Downloader, delegationTarget server.D
 	}
 
-	// Pass a list of old etags to the Downloader to prevent transfers if etags match
-	etagList := make(map[string]string)
-	for gv, specInfo := range apiService.specs {
-		etagList[gv] = specInfo.etag
-	}
-	groups, err := s.downloader.Download(apiService.handler, etagList)
+	gv, err := s.downloader.DownloadGroups(apiService.handler)
 	if err != nil {
 		return err
 	}
-
-	// Remove any groups that do not exist anymore
-	for group := range s.openAPIV3Specs[apiServiceName].specs {
-		if _, exists := groups[group]; !exists {
-			s.openAPIV3VersionedService.DeleteGroupVersion(group)
-			delete(s.openAPIV3Specs[apiServiceName].specs, group)
-		}
-	}
-
-	for group, info := range groups {
-		if info.spec == nil {
-			continue
-		}
-
-		// If ETag has not changed, no update is necessary
-		oldInfo, exists := s.openAPIV3Specs[apiServiceName].specs[group]
-		if exists && oldInfo.etag == info.etag {
-			continue
-		}
-		s.openAPIV3Specs[apiServiceName].specs[group] = &openAPIV3SpecInfo{
-			spec: info.spec,
-			etag: info.etag,
-		}
-		s.openAPIV3VersionedService.UpdateGroupVersion(group, info.spec)
-	}
+	s.apiServiceInfo[apiServiceName].gvList = gv
 	return nil
 }
 
-type specAggregator struct {
+type specProxier struct {
 	// mutex protects all members of this struct.
 	rwMutex sync.RWMutex
 
 	// OpenAPI V3 specs by APIService name
-	openAPIV3Specs map[string]*openAPIV3APIServiceInfo
-	// provided for dynamic OpenAPI spec
-	openAPIV3VersionedService *handler3.OpenAPIService
+	apiServiceInfo map[string]*openAPIV3APIServiceInfo
 
 	// For downloading the OpenAPI v3 specs from apiservices
 	downloader Downloader
 }
 
-var _ SpecAggregator = &specAggregator{}
+var _ SpecProxier = &specProxier{}
 
 type openAPIV3APIServiceInfo struct {
 	apiService v1.APIService
 	handler    http.Handler
-	specs      map[string]*openAPIV3SpecInfo
-}
-
-type openAPIV3SpecInfo struct {
-	spec *spec3.OpenAPI
-	etag string
+	gvList     []string
 }
 
 // RemoveAPIServiceSpec removes an api service from the OpenAPI map. If it does not exist, no error is returned.
 // It is thread safe.
-func (s *specAggregator) RemoveAPIServiceSpec(apiServiceName string) {
+func (s *specProxier) RemoveAPIServiceSpec(apiServiceName string) {
+	s.rwMutex.Lock()
+	defer s.rwMutex.Unlock()
+	if _, ok := s.apiServiceInfo[apiServiceName]; ok {
+		delete(s.apiServiceInfo, apiServiceName)
+	}
+}
+
+func (s *specProxier) HandleDiscovery(w http.ResponseWriter, r *http.Request) {
 	s.rwMutex.Lock()
 	defer s.rwMutex.Unlock()
-	if apiServiceInfo, ok := s.openAPIV3Specs[apiServiceName]; ok {
-		for gv := range apiServiceInfo.specs {
-			s.openAPIV3VersionedService.DeleteGroupVersion(gv)
+
+	gvList := make(map[string]bool)
+	for _, apiServiceInfo := range s.apiServiceInfo {
+		for _, gv := range apiServiceInfo.gvList {
+			gvList[gv] = true
 		}
-		delete(s.openAPIV3Specs, apiServiceName)
 	}
+
+	keys := make([]string, len(gvList))
+	i := 0
+	for k := range gvList {
+		keys[i] = k
+		i++
+	}
```

Comment:
Alternative, no strong opinion:
```suggestion
	keys := make([]string, 0, len(gvList))
	for k := range gvList {
		keys = append(keys, k)
	}
```